VU interpreter

hch12907 opened the pull request:
(Merge this only after #15 is merged)

Refactored the VU controller code (now they are split into CVu and CVuInterpreter), and (re)implemented a basic VU interpreter (though the code is certainly not as messy compared to last time ;) ).

Some fixmes (tick=fixed):
- [x] CPI is pretty much non-existent according to the interpreter :/ 
- [x] The FMAC units use the latest FDIV results in the Q register, even though the FDIV is supposedly still running, same for P register and EFU
- [x] Data hazards
- [x] No M-bit handling yet (in other words, COP2 interlock)
- [x] The priority of register writing is not considered (but right now the lower instructions execute before the upper instructions, so I guess it's partially done? eh...)
- [x] I'm kinda not satisfied with the current implementation of E-bit delay slot (this can be ignored for now)

Comments are welcomed.

hch12907 committed:
Implement VU interpreter

hch12907 committed:
Implement VU pipelines

note: this commit can be reverted if pipeline emulation is found to be unneccessary
- added MipsPipeline
- implemented VU pipelines for the interpreter

marcosatti commented on liborbum/src/Controller/Ee/Vpu/Vu/Interpreter/CVuInterpreter.cpp:
```diff
@@ -1,53 +1,246 @@
+
+        // PC & Instructions stuff...
+        const uword pc = unit->pc.read_uword() & 0x0FFF;
+        const udword raw_inst = r.ee.bus.read_uword(BusContext::Vu, 0x11000000 + 0x8000 * unit->core_id + pc);
```
This should be accessible from the vu unit's bus I think...
`core->bus.read_uword(pc);`

Check out the Resources.cpp mappings

hch12907 replied:
Tried that before but interestingly... micro mem is not mapped to the VU bus

hch12907 committed:
Implement data hazard, register writing priority, proper CPI for VU interpreter
- implemented data hazard
- implemented register writing priority
- proper CPI
- note: WAITP and WAITQ is handled as a data hazard (special case)

hch12907 committed:
fix MSVC build

hch12907 committed:
Implement M-bit handling

- only VU side, EE/COP2 side not implemented yet (another PR for that)
- this is a rough implementation, it doesnt even differentiate CPR and CCR
- also, a quick bugfix for WAITQ data hazard

hch12907 committed:
Run clang-format and update README

hch12907 commented:
Alright. This should be somewhat complete now. (Not accounting for COP2, that is)

hch12907 commented:
Closed. I will continue the development on my fork [hch12907/orbum](https://github.com/hch12907/orbum).
